onnxruntime/core/common/threadpool.cc

```diff
@@ -0,0 +1,180 @@
+      : Eigen::NonBlockingThreadPool(num_threads) {}
+
+  void ParallelFor(int32_t total, std::function<void(int32_t)> fn) {
+    // TODO: Eigen supports a more efficient ThreadPoolDevice mechanism
+    // We will simply rely on the work queue and stealing in the short term.
+    Barrier barrier(static_cast<unsigned int>(total));
+    std::function<void(int32_t)> handle_iteration;
+    handle_iteration = [=, &handle_iteration, &barrier, &fn](int iteration) {
+      fn(iteration);
+      barrier.Notify();
+    };
+
+    for (int32_t id = 0; id < total; ++id) {
+      Schedule([=, &handle_iteration]() { handle_iteration(id); });
+    }
+
+    barrier.Wait();
+  }
+
+  void ParallelForRange(int64_t first, int64_t last, std::function<void(int64_t, int64_t)> fn) {
+    // TODO: Eigen supports a more efficient ThreadPoolDevice mechanism
+    // We will simply rely on the work queue and stealing in the short term.
+    Barrier barrier(static_cast<unsigned int>(last-first+1));
+    std::function<void(int64_t, int64_t)> handle_range;
+    handle_range = [=, &handle_range, &barrier, &fn](int64_t first, int64_t last) {
+      fn(first, last);
+      barrier.Notify();
+    };
+
+    for (int64_t id = first; id <= last; ++id) {
+      Schedule([=, &handle_range]() { handle_range(id, id + 1); });
+    }
+
+    barrier.Wait();
+  }
+};
+#else
+class ThreadPool::Impl : public TaskThreadPool {
+ public:
+  Impl(const std::string& name, int num_threads)
+      : TaskThreadPool(num_threads) {}
+
+  void Schedule(std::function<void()> fn) {
+    std::packaged_task<void()> task(fn);
+    RunTask(std::move(task));
+  }
+
+  void ParallelFor(int32_t total, std::function<void(int32_t)> fn) {
+#ifdef USE_OPENMP
+#pragma omp parallel for
+    for (int32_t id = 0; id < total; ++id) {
+      fn(id);
+    }
+#else
+    for (int32_t id = 0; id < total; ++id) {
+      std::packaged_task<void()> task(std::bind(fn, id));
+      RunTask(std::move(task));
+    }
+    WaitWorkComplete();
+#endif
+  }
+
+  void ParallelForRange(int64_t first, int64_t last, std::function<void(int64_t, int64_t)> fn) {
+#ifdef USE_OPENMP
+#pragma omp parallel for
+    for (int64_t id = first; id < last; ++id) {
+      fn(id, id + 1);
+    }
+#else
+    for (int64_t id = first; id < last; ++id) {
+      std::packaged_task<void()> task(std::bind(fn, id, id + 1));
+      RunTask(std::move(task));
+    }
+    WaitWorkComplete();
+#endif
+  }
+};
+#endif
+
+//
+// ThreadPool
+//
+ThreadPool::ThreadPool(const std::string& name, int num_threads)
+    : impl_(std::make_unique<Impl>(name, num_threads)) {
+}
+
+void ThreadPool::Schedule(std::function<void()> fn) { impl_->Schedule(fn); }
+
+void ThreadPool::ParallelFor(int32_t total, std::function<void(int32_t)> fn) { 
+  if (total <= 0) {
+    return;
+  }
+
+  impl_->ParallelFor(total, fn);
+}
+
+void ThreadPool::ParallelForRange(int64_t first, int64_t last, std::function<void(int64_t, int64_t)> fn) { 
+  if (last <= first) {
```

Comment:
I have a doubt. If (last < first) should we call the function?

Reply:
We don't control the semantics of the function to be invoked.  This is an assertion that if the invocation does not follow the API pattern, we will defer to the function, not fail it and not parallelize it.

Reply:
understood. Thanks.